### PR TITLE
[CARBONDATA-3102] Fix NoClassDefFoundError when use thriftServer and beeline to read/write data from/to S3

### PR DESCRIPTION
**Pull request by xubo245**
When I run the carbonThriftServer and use beeline to read/write data from/to S3, it throw  NoClassDefFoundError:
`0: jdbc:hive2://127.0.0.1:10000> create table xubo6 stored by 'carbondata' location 's3a://xubo6/xubo3';
Error: java.lang.NoClassDefFoundError: org/apache/http/pool/ConnPoolControl (state=,code=0)`

Reproduce step in local:

1. start carbonThriftServer with s3 store path, ak,sk,endpoint.  

2.use beeline to connect:
` localhost:spark xubo$ ./bin/beeline -u jdbc:hive2://127.0.0.1:10000`

3. run sql
`create table xubo6 stored by 'carbondata' location 's3a://xubo6/xubo3';`
Exception:
`
0: jdbc:hive2://127.0.0.1:10000> create table xubo6 stored by 'carbondata' location 's3a://xubo6/xubo3';
Error: java.lang.NoClassDefFoundError: org/apache/http/pool/ConnPoolControl (state=,code=0)
`

 When I run in cluster, also has this problem before.

This PR fix NoClassDefFoundError when use thriftServer and beeline 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 No
 - [x] Any backward compatibility impacted?
 No
 - [x] Document update required?
No
 - [x] Testing done
    No
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
No

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1423/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9680/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1632/

**Comment by xubo245** [comment]
@jackylk @KanakaKumar @kunal642 Please review it.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1429/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1638/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9686/

**Comment by kunal642** [comment]
Please add description

**Comment by kunal642** [comment]
@xubo245 I what scenario are you facing this issue? I am not facing this issue.

File: integration/spark2/pom.xml [hunk 1]
```diff
@@ -134,6 +134,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
```
Comment (kunal642) [inline integration/spark2/pom.xml]: Why is this dependency needed in spark2?

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala [hunk 2]
```diff
@@ -48,8 +48,13 @@ object CarbonThriftServer {
       System.exit(0)
     }
 
+    val master = Option(System.getProperty("spark.master"))
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala]: Was this the reason for NoClassDefFoundError?

File: integration/spark2/pom.xml (hunk 1 above)
Comment (xubo245) [reply integration/spark2/pom.xml]: Because when I use org.apache.carbondata.spark.thriftserver.CarbonThriftServe and use beeline to connect, it throw NoClassDefFoundError。NoClassDefFoundError: org/apache/http/impl/conn/PoolingClientConnectionManager。  PoolingClientConnectionManager is belong to httpclient, so I add:

    <dependency>
      <groupId>org.apache.httpcomponents</groupId>
      <artifactId>httpclient</artifactId>
      <version>${httpclient.version}</version>
    </dependency>

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala (hunk 2 above)
Comment (xubo245) [reply integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala]: No, it's support CarbonThriftServer run in local/cluster, which convenient for users. Before if user want to run CarbonThriftServer in local, user need add master("local")

File: integration/spark2/pom.xml (hunk 1 above)
Comment (kunal642) [reply integration/spark2/pom.xml]: But nobody else is facing this issue. And also none of our environments are getting this exception. Maybe u need to add httpclient jars to you cluster seperately?

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala (hunk 2 above)
Comment (kunal642) [reply integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala]: ok

File: integration/spark2/pom.xml (hunk 1 above)
Comment (xubo245) [reply integration/spark2/pom.xml]: I meet more than two times. This issue also happen in spark on k8s before. You can try in original cluster. 
It's not convenient for user to  add httpclient jars to you cluster seperately.  What's effect if we add this dependency?

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 3]
```diff
@@ -157,8 +157,8 @@ object S3Example {
   }
 
   def getSparkMaster(args: Array[String]): String = {
-      if (args.length == 5) args(4)
-      else if (args(3).contains("spark:") || args(3).contains("mesos:")) args(3)
-      else "local"
-    }
```
Comment (akashrn5) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: revert the unnecessary changes

Comment (xubo245) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: the format is incorrect. we should optimize it.
```def getS3EndPoint(args: Array[String]): String = {		   def getS3EndPoint(args: Array[String]): String = {
     if (args.length >= 4 && args(3).contains(".com")) args(3)		     if (args.length >= 4 && args(3).contains(".com")) args(3)
     else ""		     else ""
   }```

**Comment by kunal642** [comment]
@xubo245 I dont think we are supporting thriftServer to be run through IntelliJ on local. In cluster mode this jar would already be available as part of spark dependency. 
For SDK because it is dependent on all modules therefore httpclient is required.

@manishgupta88 @kumarvishal09 @jackylk @ravipesala What are your opinions?

**Comment by xubo245** [comment]
@kunal642 In cluster, it also has this problem.

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala (hunk 2 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala]: I think better to let user give --master local, user can control this.

File: integration/spark2/pom.xml (hunk 1 above)
Comment (jackylk) [reply integration/spark2/pom.xml]: I have never encounter this problem before, can you describe the scenario in more detail, in what case NoClassDefFoundError will be thrown?

Comment (jackylk) [reply integration/spark2/pom.xml]: I see `org.apache.httpcomponents` is already introduced in other modules, please unify the version for this jar in all places

Comment (xubo245) [reply integration/spark2/pom.xml]: ok, done

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala (hunk 2 above)
Comment (xubo245) [reply integration/spark2/src/main/scala/org/apache/carbondata/spark/thriftserver/CarbonThriftServer.scala]: ok, done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10030/

**Comment by jackylk** [comment]
LGTM
